### DISCUSSION TITLE
General feedback

### DISCUSSION THREAD

**Discussion by jumarko** [jacekschae/learn-pedestal-course-files#19] [Feedback]
Once again, thanks for the great course and lots of new videos recently!

I want to mention a couple of "problems" I've had with the setup and the presentation method: 

## Dev workflow
As I said before, I found the development workflow problematic - the handler functions seem to be captured somehow and an updated definition is not available without server restart although [I reference handlers as vars](https://github.com/jumarko/learn-pedestal/pull/13/files#diff-93fcb18e3b4a9aa3ad371bdc883cd94a1dcb6f5f7e794ed765e7490a990d0460R17) (via `#'`)

We discussed this earlier and I heard that there will ultimately be a better workflow but I think it would be better to introduce this earlier, perhaps as an extra lesson after the last one in the Component series (Component REPL)

## Wanted: Smaller steps

You often write way too much code, for my taste, in one go

Example: in lesson 26 Retrieve recipe, you wrote the whole retrieve-recipe-response thing without testing it. You even mentioned it briefly at the end that you think it's fine to go without testing incrementally. I don't think so.

I'd much prefer to take smaller steps, especially since this is all about learning. 
So I'd perhaps wrote this first: https://github.com/jumarko/learn-pedestal/pull/13/commits/4fd4d616c539eadfdff5b1440138a5d8ebbb855f

(defn retrieve-recipe-response
  [{:keys [path-params system/database] :as request}]
  (let [db (:db database)
        recipe-id (parse-uuid (:recipe-id path-params))]
    (response/response (str recipe-id))))
  
Then test it: 
```
  (pt/response-for (-> cr/system :api-server :service ::http/service-fn)
                   :get recipe-url
                   :headers {"Authorization"  "auth|5fbf7db6271d5e0076903601"})
;; => {:status 200, :body "7e485d7a-23c1-4519-8043-15dbb38775b7", ...
```

Then I'd implement the query part: https://github.com/jumarko/learn-pedestal/pull/13/commits/51c533a4ceccb3f7cee2c21d16259ea023faf826
```
(defn- retrieve-recipe [db recipe-id]
  (let [recipe (d/q '[:find (pull ?e pattern)
                      :in $ ?recipe-id pattern
                      :where [?e :recipe/recipe-id ?recipe-id]]
                    db recipe-id recipe-pattern)]
    recipe))

(defn retrieve-recipe-response
  ;; Notice `path-params`
  [{:keys [path-params system/database] :as _request}]
  (let [db (:db database)
        recipe-id (parse-uuid (:recipe-id path-params))]
    (response/response (retrieve-recipe db recipe-id))))
```

Again, these are little annoyances - the course is great anyway and thank you again for all the hard work.

**Comment by jacekschae** [comment]
Thanks so much for your comments. I should have done a better job explaining why I went this route -- skipping tests and repl -- it's much better explained toward the end of video 31 - Retract recipe. There is a lot going on in the Recipes videos and I focus only on the flow with interceptors without testing and repl. 

I'll take your pointes into the account going forward and will try to do a better job with this. I hope you will have a chance to watch video 31 and let me know what you think about it.

**Reply by jumarko** [reply]
I'll definitely watch it.
Thanks for the reply!
